source/grid/tria_accessor.cc

```diff
@@ -1329,139 +1329,53 @@ namespace
-                       0.61419262306231814 * f) +
-                   0.056712962962962937 *
-                     std::sqrt(a + 0.86595709258347853 * b +
-                               0.44888729929169013 * c +
-                               1.8611363115940525 * d + 1.3399810435848563 * e +
-                               1.2469436885317342 * f) +
-                   0.03025074832140047 *
-                     std::sqrt(a + 0.86595709258347853 * b +
-                               0.86595709258347853 * c +
-                               1.8611363115940525 * d + 1.8611363115940525 * e +
-                               1.7319141851669571 * f);
-          }
-
-        // the face is planar. then its area is 1/2 of the norm of the
-        // cross product of the two diagonals
-        const Tensor<1, 3> v12        = accessor.vertex(2) - accessor.vertex(1);
-        const Tensor<1, 3> twice_area = cross_product_3d(v03, v12);
-        return 0.5 * twice_area.norm();
+        const Point<3> x0 = accessor.vertex(0), x1 = accessor.vertex(1),
+                       x2 = accessor.vertex(2), x3 = accessor.vertex(3);
+
+        // This is based on the approach used in libMesh (see face_quad4.C): the
+        // primary differences are the vertex numbering and quadrature order.
+        //
+        // The area of a surface is the integral of the magnitude of its normal
+        // vector, which may be computed via the cross product of two tangent
+        // vectors. We can easily get tangent vectors from the surface
+        // parameterization. Hence, given a bilinear surface
+        //
+        //     P(chi, eta) = x0 + (x1 - x0) chi + (x2 - x0) eta
+        //                 + (x3 + x0 - x1 - x2) chi eta
+        //
+        // the tangent vectors are
+        //
+        //     t1 = (x1 - x0) + (x3 + x0 - x1 - x2) eta
+        //     t2 = (x2 - x0) + (x3 + x0 - x1 - x2) xi
+        const Tensor<1, 3> a = (x3 + x0 - x1 - x2), b0 = x1 - x0, b1 = x2 - x0;
+
+        // If we have a parallelogram then return quickly.
+        if (a.norm_square() < 1e-14 * accessor.diameter())
+          return cross_product_3d(b0, b1).norm();
+
+        // Otherwise, use a 4x4 quadrature to approximate the surface area.
+        // Hard-code this in to prevent the extra overhead of always creating
+        // the same QGauss rule.
+        constexpr unsigned int n_qp = 4;
+        const double           c1   = 2.0 / 7.0 * std::sqrt(6.0 / 5.0);
+        const double           w0   = (18.0 - std::sqrt(30)) / 72.0;
+        const double           w1   = (18.0 + std::sqrt(30)) / 72.0;
+
+        const std::array<double, n_qp> q{{
+          0.5 - std::sqrt(3.0 / 7.0 + c1) / 2.0,
+          0.5 - std::sqrt(3.0 / 7.0 - c1) / 2.0,
+          0.5 + std::sqrt(3.0 / 7.0 - c1) / 2.0,
+          0.5 + std::sqrt(3.0 / 7.0 + c1) / 2.0,
+        }};
+        const std::array<double, n_qp> w{{w0, w1, w1, w0}};
```

Comment:
These can be also be magic numbers and turned into `constexpr`.

Reply:
I'd like to mark these as `constexpr` but `std::sqrt()` will not be `constexpr` until C++26. Like Martin said I don't think it makes a difference (and GCC already has a `constexpr` version of square root - it looks like clang applies the same optimization in a different way)